keps/sig-network/3836-kube-proxy-improved-ingress-connectivity-reliability/README.md

```diff
@@ -0,0 +1,346 @@
+```
+
+## Motivation
+
+The motivation for each change is:
+
+1. Not including the `healthz` state can cause Kube-proxy to indicate to a load
+  balancer that it should send traffic to a Node simply because the endpoint is
+  scheduled there, even though Kube-proxy might not be healthy and successfully
+  managed to write the rules required for actually being able to forward traffic
+  to the endpoint.
+
+2. Workloads on terminating Nodes would benefit from connection draining as to
+  have their connections gracefully terminated. Workloads running on terminating
+  nodes get drained simultaneously with Kube-proxy today, this means that
+  there's no guarantee that connection draining will be enabled for them
+  currently. That might be the case if Kube-proxy is terminated before the
+  endpoint, but there's no guarantee of that. If Kube-proxy reacts to a "node is
+  about to go away" signal, then Kube-proxy can start prematurely failing the HC
+  and enable connection draining for these workloads.
+
+### Goals
+
+- Have kube-proxy accurately reflect its `healthz` state in its response to LB HCs
+  for class B) services.
+
+- Connection drain terminating Nodes
+
+### Non-Goals
+
+- Aligning cloud provider HCs for class A) services. Cloud providers like
+  Azure/AWS do not configure their HCs to point to the service proxy. Instead
+  they connect to the `NodePort` defined for the service. As to have them
+  benefit from the proposals of this KEP, they would need to change their
+  implementation.
+
+## Proposal
+
+#### Risk
+
+The risk with each proposal is:
+
+1. None. Class B) type services which benefitted from being scheduled on nodes
+   with a non-healthy Kube-proxy but which had successfully written the rules
+   required for traffic to reach the endpoint: will not benefit from that
+   anymore. But that's not deterministic...and that's not really a "risk".
+
+2. If the right field indicating "node is terminating / deleting" is
+   chosen/exists: None
+
+#### Mitigations
+
+None, since there are no Risks.
+
+## Design Details
+
+For:
+
+1. Implement change in Kube-proxy which allows the healthcheck server access to
+   the `healthz` state.
+
+2. Possibly implement a Node watcher (I believe it already exists) in
+   Kube-proxy. I also believe there was a discussion about removing said watcher
+   for other reasons, but I am not sure where the consensus lies on that?
+
+### Test Plan
+
+#### Prerequisite testing updates
+
+#### Unit tests
+
+Update the Kube-proxy unit tests to include the `healthz` answer for the
+healthcheck server test suite.
+
+#### Integration tests
+
+Kubernetes is mostly tested via unit tests and e2e, not integration, and this is
+not expected to change.
+
+#### e2e tests
+
+- Add E2E tests for connectivity to services on terminating nodes and
+  validate graceful termination of TCP connections.
+
+### Graduation Criteria
+
+#### GA
+
+This is addressing a sub-optimal solution currently existing in Kubernetes, and
```

Comment:
I don't like going directly to GA until we have tested and verified this new approach, I strongly suggest to go with alpha, beta and GA stages to be on the safe said

Reply:
oh, agree - I missed this.  This one probably deserves an Alpha

Reply:
OK, putting alpha